### PR TITLE
Add missing error messages to asserts in QuotesImpl

### LINKED ISSUE #20946: All `assert`s in `Quotes` should give helpful error messages
## Compiler version

`main` (`3.5.1-RC1-bin-20240628-1efbb92-NIGHTLY`)

## Minimized example

```scala
//> using scala 3.5.1-RC1-bin-20240628-1efbb92-NIGHTLY
// _2.scala

inline def macroWithAssertFailing[T](t: T): Unit = ${ macroWithAssertFailingImpl[T]('t) }

@main
def run =
  macroWithAssertFailing[Int](123)
```

Comment out the `Ref` expression and uncomment the `TypeIdent` one for the second error.

```scala
// _1.scala

import scala.quoted.*

def macroWithAssertFailingImpl[T: Type](t: Expr[T])(using Quotes): Expr[Unit] = {
  import quotes.reflect.*

  Ref(TypeRepr.of[T].typeSymbol) // error 1

  // TypeIdent(t.asTerm.symbol) // error 2

  '{()}
}
```

#### Error 1

```scala
-- Error: macroasserts/_2.scala:8:29 ----------------------------
8 |  macroWithAssertFailing[Int](123)
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |Exception occurred while executing macro expansion.
  |java.lang.AssertionError: assertion failed
  |	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
  |	at scala.quoted.runtime.impl.QuotesImpl$reflect$Ref$.apply(QuotesImpl.scala:475)
  |	at scala.quoted.runtime.impl.QuotesImpl$reflect$Ref$.apply(QuotesImpl.scala:474)
  |	at _1$package$.macroWithAssertFailingImpl(_1.scala:8)
  |
  |-----------------------------------------------------------------------------
  |Inline stack trace
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  |This location contains code that was inlined from _2.scala:4
4 |inline def macroWithAssertFailing[T](t: T): Unit = ${ macroWithAssertFailingImpl[T]('t) }
  |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   -----------------------------------------------------------------------------
```

#### Error 2

```scala
-- Error: macroasserts/_2.scala:8:29 ----------------------------
8 |  macroWithAssertFailing[Int](123)
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |Exception occurred while executing macro expansion.
  |java.lang.AssertionError: assertion failed
  |	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
  |	at scala.quoted.runtime.impl.QuotesImpl$reflect$TypeIdent$.apply(QuotesImpl.scala:1165)
  |	at scala.quoted.runtime.impl.QuotesImpl$reflect$TypeIdent$.apply(QuotesImpl.scala:1164)
  |	at _1$package$.macroWithAssertFailingImpl(_1.scala:10)
  |
  |-----------------------------------------------------------------------------
  |Inline stack trace
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  |This location contains code that was inlined from _2.scala:4
4 |inline def macroWithAssertFailing[T](t: T): Unit = ${ macroWithAssertFailingImpl[T]('t) }
  |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   -----------------------------------------------------------------------------
```

## Why this Error/Warning was not helpful

The assert should provide a useful message about why it failed.

Asserts in `Quotes` that are missing messages:
- https://github.com/scala/scala3/blob/0e60d36e80522a60c1ab1d3a686ff2e5194acfd1/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala#L475
- https://github.com/scala/scala3/blob/0e60d36e80522a60c1ab1d3a686ff2e5194acfd1/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala#L1165

It might also be useful to check other assert messages if they can be improved.

### PR DESCRIPTION
**Pull request by KacperFKorban**
closes #20946

### CODE REVIEW

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 1]
```diff
@@ -472,7 +472,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
       def term(tp: TermRef): Ref =
         withDefaultPos(tpd.ref(tp).asInstanceOf[tpd.RefTree])
       def apply(sym: Symbol): Ref =
-        assert(sym.isTerm)
+        assert(sym.isTerm, s"expected a term symbol but received $sym")
```
Comment (jchyb) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:475]:
```suggestion
        assert(sym.isTerm, s"expected a term symbol, but received $sym")
```

**Review by jchyb — COMMENTED** [review]
The changes itself LGTM! Just found the smallest typo. I do have one concern for the tests though... we have two of these already, and the visible stack trace makes it require regenerating check files with every change to QuotesImpl.scala (I do intend to adjust those, one day...). Perhaps we could catch the assertion error in the macro, check the message, rethrow it, and not have a checkfile here?

**Comment by KacperFKorban** [comment]
Right! I didn't think about the fact that the whole stack trace is very implementation dependent. Looks a bit hacky now, but works. Nice catch!

**Review by jchyb — APPROVED** [review]
Thank you!

**Pull request merged**